source/core/NumericString.cpp

```diff
@@ -1953,6 +1954,97 @@ VIREO_FUNCTION_SIGNATUREV(StringScan, StringScanParamBlock)
+    _ParamDef(ErrorCluster, ErrorInOut);
+    NEXT_INSTRUCTION_METHOD()
+};
+
+static bool IsSupportedEncodingAndUpdateErrorAndMakeOutputArrayEmpty(ErrorCluster* errorCluster, Int32 stringEncoding, TypedArrayCoreRef outputArray) {
+    if (stringEncoding != 0) {
+        if (errorCluster != nullptr) {
+            errorCluster->SetError(true, -1, "Encoding type not supported", true);  // TODO(sanmut): Update error code.
+        }
+        if (outputArray != nullptr)
+            outputArray->Resize1D(0);
+        return false;
+    }
+    return true;
+}
+
+static bool IsValidUTF8AndUpdateErrorAndMakeOutputArrayEmpty(UInt8* begin, UInt8* end, ErrorCluster* errorCluster, TypedArrayCoreRef outputArray)
+{
+    if (!utf8::is_valid(begin, end)) {
+        if (errorCluster != nullptr) {
+            errorCluster->SetError(true, 1, "Invalid UTF-8 characters found");  // TODO(sanmut): Update error code.
+        }
+        if (outputArray != nullptr)
+            outputArray->Resize1D(0);
+        return false;
+    }
+    return true;
+}
+
+VIREO_FUNCTION_SIGNATURET(ByteArrayToString, ByteArrayToStringParamBlock)
+{
+    StringRef stringOut = _ParamPointer(StringOut) ? _Param(StringOut) : nullptr;
+    ErrorCluster *errorCluster = _ParamPointer(ErrorInOut);
+    if ((stringOut == nullptr && errorCluster == nullptr) || (errorCluster != nullptr && errorCluster->hasError()))
+        return _NextInstruction();
+    TypedArrayCoreRef byteArray = _Param(ByteArrayIn);
+    TypeRef eltType = byteArray->ElementType();
+    VIREO_ASSERT(eltType->IsA(&TypeCommon::TypeInt8) || eltType->IsA(&TypeCommon::TypeUInt8));
+    IntIndex rank = byteArray->Rank();
+    VIREO_ASSERT(rank == 1);
+    Int32 stringEncoding = _Param(StringEncodingIn);
+    Int32 arrayLength = byteArray->DimensionLengths()[0];
+    InstructionCore* value;
+    if (!IsSupportedEncodingAndUpdateErrorAndMakeOutputArrayEmpty(errorCluster, stringEncoding, stringOut)) {
+        return _NextInstruction();
+    }
+    if (!IsValidUTF8AndUpdateErrorAndMakeOutputArrayEmpty(byteArray->RawBegin(), byteArray->RawBegin() + arrayLength, errorCluster, stringOut)) {
+        return _NextInstruction();
+    }
+    stringOut->Replace1D(0, arrayLength, byteArray->RawBegin(), true);
+    return _NextInstruction();
+}
+
+struct StringToByteArrayParamBlock : InstructionCore
+{
+    _ParamDef(TypedArrayCoreRef, ByteArrayOut);
+    _ParamDef(StringRef, StringIn);
+    _ParamDef(Int32, StringEncodingIn);  // 0 = UTF-8 (only supported)
+    _ParamDef(ErrorCluster, ErrorInOut);
+    NEXT_INSTRUCTION_METHOD()
+};
+
+VIREO_FUNCTION_SIGNATURET(StringToByteArray, StringToByteArrayParamBlock)
+{
+    StringRef stringIn = _Param(StringIn);
+    TypedArrayCoreRef byteArrayOut = _ParamPointer(ByteArrayOut) ? _Param(ByteArrayOut) : null;
+    ErrorCluster *errorCluster = _ParamPointer(ErrorInOut);
+    if (byteArrayOut == nullptr || (errorCluster != nullptr && errorCluster->hasError()))
```

Comment:
In the other function you are checking if the output array is null AND the error cluster is null. Please do the same here.

Reply:
No, this is correct here. These are the conditions where we should do nothing and return early.

Reply:
Just to be clear, I was asserting that the check should be (byteArrayOut == nullptr && errorCluster == nullptr) || (errorCluster != nullptr && errorCluster->hasError()).

Reply:
Addressed this.